From 37c198d8f856736d6cdde079de8c77a48fa4a6fe Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:17:48 -0400
Subject: [net] netfilter: nf_tables: fix missing return trace at the end of non-base chain

Message-id: <20141021161748.22888.2058.stgit@localhost.localdomain>
Patchwork-id: 97944
O-Subject: [RHEL7 net PATCH 047/124] netfilter: nf_tables: fix missing return trace at the end of non-base chain
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: 7e9bc10db275b22a9db0f976b33b5aeed288da73

 commit 7e9bc10db275b22a9db0f976b33b5aeed288da73
 Author: Pablo Neira Ayuso <pablo@netfilter.org>
 Date:   Sun May 11 17:14:49 2014 +0200

    netfilter: nf_tables: fix missing return trace at the end of non-base chain

    Display "return" for implicit rule at the end of a non-base chain,
    instead of when popping chain from the stack.

    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nf_tables_core.c b/net/netfilter/nf_tables_core.c
index 711c7f6..4e6e681 100644
--- a/net/netfilter/nf_tables_core.c
+++ b/net/netfilter/nf_tables_core.c
@@ -182,18 +182,16 @@ next_rule:
  case NFT_RETURN:
   if (unlikely(pkt->skb->nf_trace))
    nft_trace_packet(pkt, chain, rulenum, NFT_TRACE_RETURN);
-
-  /* fall through */
+  break;
  case NFT_CONTINUE:
+  if (unlikely(pkt->skb->nf_trace && !(chain->flags & NFT_BASE_CHAIN)))
+   nft_trace_packet(pkt, chain, ++rulenum, NFT_TRACE_RETURN);
   break;
  default:
   WARN_ON(1);
  }
 
  if (stackptr > 0) {
-  if (unlikely(pkt->skb->nf_trace))
-   nft_trace_packet(pkt, chain, ++rulenum, NFT_TRACE_RETURN);
-
   stackptr--;
   chain = jumpstack[stackptr].chain;
   rule  = jumpstack[stackptr].rule;
-- 
1.7.1